agent/proxycfg/state.go

```diff
@@ -965,6 +1025,241 @@ func (s *state) handleUpdateMeshGateway(u cache.UpdateEvent, snap *ConfigSnapsho
+		}
+		watchedSvcs[u.Identifier()] = struct{}{}
+	}
+
+	for id, cancelFn := range snap.IngressGateway.WatchedDiscoveryChains {
+		if _, ok := watchedSvcs[id]; !ok {
+			cancelFn()
+			delete(snap.IngressGateway.WatchedDiscoveryChains, id)
+		}
+	}
+
+	return nil
+}
+
+func upstreamForIngress(address, namespace, service string, port int) structs.Upstream {
+	if address == "" {
+		address = "0.0.0.0"
+	}
+	return structs.Upstream{
+		DestinationName:      service,
+		DestinationNamespace: namespace,
+		LocalBindAddress:     address,
+		LocalBindPort:        port,
+	}
+}
+
+func (s *state) watchIngressDiscoveryChain(snap *ConfigSnapshot, u structs.Upstream) error {
+	if _, ok := snap.IngressGateway.WatchedDiscoveryChains[u.Identifier()]; ok {
+		return nil
+	}
+
+	ctx, cancel := context.WithCancel(s.ctx)
+	err := s.cache.Notify(ctx, cachetype.CompiledDiscoveryChainName, &structs.DiscoveryChainRequest{
+		Datacenter:           s.source.Datacenter,
+		QueryOptions:         structs.QueryOptions{Token: s.token},
+		Name:                 u.DestinationName,
+		EvaluateInDatacenter: s.source.Datacenter,
+		EvaluateInNamespace:  u.DestinationNamespace,
+		// OverrideMeshGateway:    s.proxyCfg.MeshGateway.OverlayWith(u.MeshGateway),
+		// OverrideConnectTimeout: cfg.ConnectTimeout(),
+	}, "discovery-chain:"+u.Identifier(), s.ch)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	snap.IngressGateway.WatchedDiscoveryChains[u.Identifier()] = cancel
+	return nil
+}
+
+func (s *state) resetIngressWatchesFromChain(
+	id string,
+	chain *structs.CompiledDiscoveryChain,
+	snap *ConfigSnapshot,
+) error {
+	s.logger.Trace("resetting watches for discovery chain", "id", id)
+	if chain == nil {
+		return fmt.Errorf("not possible to arrive here with no discovery chain")
+	}
+
+	// Initialize relevant sub maps.
+	if _, ok := snap.IngressGateway.WatchedUpstreams[id]; !ok {
+		snap.IngressGateway.WatchedUpstreams[id] = make(map[string]context.CancelFunc)
+	}
+	if _, ok := snap.IngressGateway.WatchedUpstreamEndpoints[id]; !ok {
+		snap.IngressGateway.WatchedUpstreamEndpoints[id] = make(map[string]structs.CheckServiceNodes)
+	}
+
+	// We could invalidate this selectively based on a hash of the relevant
+	// resolver information, but for now just reset anything about this
+	// upstream when the chain changes in any way.
+	for targetID, cancelFn := range snap.IngressGateway.WatchedUpstreams[id] {
+		s.logger.Trace("stopping watch of target",
+			"upstream", id,
+			"chain", chain.ServiceName,
+			"target", targetID,
+		)
+		delete(snap.IngressGateway.WatchedUpstreams[id], targetID)
+		delete(snap.IngressGateway.WatchedUpstreamEndpoints[id], targetID)
+		cancelFn()
+	}
+
+	for _, target := range chain.Targets {
+		s.logger.Trace("initializing watch of target",
+			"upstream", id,
+			"chain", chain.ServiceName,
+			"target", target.ID,
+			"mesh-gateway-mode", target.MeshGateway.Mode,
+		)
+
+		ctx, cancel := context.WithCancel(s.ctx)
+		err := s.watchConnectProxyService(
```

Comment:
The PR description alludes to this, but I get the suspicion that there's a bunch of missing plumbing for basically replicating the bulk of the `connect-proxy` workflow for ingress gateways.

The more you duplicate it, the more it might make sense to somehow rearrange the snapshot state so that there's a common "handle" to a collection of stuff related to resolving a discovery chain and all of the follow-on stuff after that, so you don't end up in hundreds of lines of copypasta.

For example, I think you event switch loop would have to take into account having the *ingress gateway* listen for *mesh gateways* as a side effect of evaluating a discovery chain.

Reply:
Exactly right. The `proxycfg` code in particular ended up mostly duplicated so that we could reason our way through the code more effectively.

I am not quite sure what a good rearrangement of the snapshot state might look like here. I actually like the clear delineation of code paths presented by lines such as:
```
	switch s.kind {
	case structs.ServiceKindConnectProxy:
		return s.initWatchesConnectProxy()
	case structs.ServiceKindMeshGateway:
		return s.initWatchesMeshGateway()
	case structs.ServiceKindIngressGateway:
		return s.initWatchesIngressGateway()
	default:
		return fmt.Errorf("Unsupported service kind")
	}
```

But ideally would be able to reuse a bunch of logic internally.

Reply:
Yeah I like having the top level `<something><proxy-type>()` calls. Currently the proxycfg and xds packages all follow that calling pattern. The tricky part is that the `Snapshot` is similarly partitioned.

Maybe you could do something like this (strawman for comment brevity):
```
type Snapshot struct {
  ConnectProxy ConnectProxySnap
  MeshGateway MeshGatewaySnap
  IngressGateway IngressGatewaySnap
}

type ConnectProxySnap struct { ... fields ... }
type MeshGatewaySnap struct { ... fields ... }
type IngressGatewaySnap struct { ... fields ... }

var (
  _ SnapshotWithUpstreams = (*ConnectProxySnap)(nil)
  _ SnapshotWithUpstreams = (*MeshGatewaySnap)(nil)
  _ SnapshotWithUpstreams = (*IngressGatewaySnap)(nil)
)
 
// SnapshotWithUpstreams represents getter/setter operations against
// the proxy-specific snapshot state so you can make generic functions
// for these sorts of shared operations rather than the ones that currently
// just take a Snapshot
type SnapshotWithUpstreams interface {
  UpsertUpstreamDiscoveryChain(...args...)
  DeleteUpstreamDiscoveryChain(...args...)
  UpsertUpstreamEndpoints(...args...)
  DeleteUpstreamEndpoints(...args...)
  // .. more ...
}
```

Alternatively you could come up with a struct like `CommonUpstreamSnapshot` that IngressGateway and ConnectProxy line items both embed, but I don't know enough about the way you hope to structure the ingress proxy to know if you are going to _store_ them slightly differently for reasons.

Reply:
Ok, we refactored this to pull out a struct `ConfigSnapshotUpstreams` that is embedded in both the `configSnapshotConnectProxy` and `configSnapshotIngressGateway` types. This allowed us to reuse much of the update handling logic by pulling out a `handleUpdateUpstreams` function that just takes in the shared data structure.